transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringDatagramChannel.java

```diff
@@ -438,11 +426,46 @@ protected void readComplete0(int res) {
                     allocHandle.readComplete();
                     pipeline.fireChannelReadComplete();
                 }
-            } catch (Throwable t) {
-                if (connected && t instanceof NativeIoException) {
-                    t = translateForConnected((NativeIoException) t);
+            } finally {
+                if (byteBuf != null) {
+                    byteBuf.release();
+                }
+            }
+        }
+
+        private void recvmsgComplete(ChannelPipeline pipeline, IOUringRecvByteAllocatorHandle allocHandle,
+                                      ByteBuf byteBuf, int res, int idx, int outstanding) throws IOException {
+            try {
+                MsgHdrMemory hdr = recvmsgHdrs.hdr(idx);
+
+                if (res < 0) {
+                    // If res is negative we should pass it to ioResult(...) which will either throw
+                    // or convert it to 0 if we could not read because the socket was not readable.
+                    allocHandle.lastBytesRead(ioResult("io_uring recvmsg", res));
+                } else if (res > 0) {
+                    allocHandle.lastBytesRead(res);
+                    allocHandle.incMessagesRead(1);
+                    DatagramPacket packet = hdr.read(IOUringDatagramChannel.this, byteBuf, res);
+                    pipeline.fireChannelRead(packet);
+                } else {
+                    allocHandle.lastBytesRead(0);
+                }
+                byteBuf = null;
```

Comment:
Our finally-clause only releases the buffer if `ioResult` throws? Maybe this can be structured better?

Reply:
I think @njhill is right here...